src/abTestingInfra/Experiment/index.tsx

```diff
@@ -0,0 +1,133 @@
+import React from "react";
+import { assertNameIsValid } from "../utils";
+import { ABTestingContextConsumer } from "../ABTestingContext";
+import PropTypes from "prop-types";
+import { recordPlay } from "../emitter";
+import { ABTestingPayload, VariantProps, Variants } from "../types";
+import ExperimentContext from "../ExperimentContext";
+
+class Experiment extends React.Component {
+  static propTypes = {
+    name: PropTypes.string.isRequired,
+    defaultVariantName: PropTypes.string.isRequired,
+    children: PropTypes.node.isRequired,
+    triggerPlay: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
+  };
+
+  props: {
+    name: string;
+    defaultVariantName: string;
+    triggerPlay?: (() => void) | Promise<void>;
+    children: React.ReactNode;
+  };
+
+  variants: Variants;
+  variantName: string;
+
+  constructor(props) {
+    super(props);
+
+    this.variants = Experiment.filterVariants(props.children);
+    this.variantName = "undefined";
+  }
+
+  componentDidMount(): void {
+    const { name, triggerPlay } = this.props;
+
+    if (typeof window === "undefined") {
+      return;
+    }
+
+    if (!!triggerPlay && typeof triggerPlay === "function") {
+      Promise.resolve(triggerPlay())
+        .then(() => recordPlay(name, this.variantName))
+        .catch((err) =>
+          console.error(
+            `triggerPlay hook throw an error. Error: ${err.toString()}`
+          )
+        );
+
+      return;
+    }
+
+    recordPlay(name, this.variantName);
+  }
+
+  static filterVariants(children: React.ReactNode): Variants {
+    const variants: Variants = {};
+
+    React.Children.forEach(
```

Comment:
Same as above — if we need these, please `import { Children, Component, SFC, ReactNode } from 'react';`

Reply:
Done